subsys/bluetooth/services/ots_oacp.c

```diff
@@ -0,0 +1,339 @@
+		break;
+	case BT_GATT_OTS_OACP_PROC_ABORT:
+		break;
+	default:
+		return true;
+	}
+
+	return (len == ref_len);
+}
+
+static void oacp_read_proc_cb(struct bt_gatt_ots_l2cap *l2cap_ctx,
+			      struct bt_conn *conn)
+{
+	int err;
+	uint8_t *obj_chunk;
+	uint32_t offset;
+	uint32_t len;
+	struct bt_gatt_ots *ots;
+	struct bt_gatt_ots_object_read_op *read_op;
+
+	ots     = CONTAINER_OF(l2cap_ctx, struct bt_gatt_ots, l2cap);
+	read_op = &ots->cur_obj->state.read_op;
+	offset  = read_op->oacp_params.offset + read_op->sent_len;
+
+	if (read_op->sent_len >= read_op->oacp_params.len) {
+		LOG_DBG("OACP Read Op over L2CAP is completed");
+
+		if (read_op->sent_len > read_op->oacp_params.len) {
+			LOG_ERR("More bytes sent that the client requested");
+		}
+
+		ots->cur_obj->state.type = BT_GATT_OTS_OBJECT_IDLE_STATE;
+		ots->cb->obj_read(ots, conn, ots->cur_obj->metadata.id, NULL,
+				  0, offset);
+		return;
+	}
+
+	len = read_op->oacp_params.len - read_op->sent_len;
+	len = ots->cb->obj_read(ots, conn, ots->cur_obj->metadata.id,
+			       &obj_chunk, len, offset);
+
+	ots->l2cap.tx_done = oacp_read_proc_cb;
+	err = bt_gatt_ots_l2cap_send(&ots->l2cap, obj_chunk, len);
+	if (err) {
+		LOG_ERR("L2CAP CoC error: %d while trying to execute OACP "
+			"Read procedure", err);
+		ots->cur_obj->state.type = BT_GATT_OTS_OBJECT_IDLE_STATE;
+	} else {
+		read_op->sent_len += len;
+	}
+}
+
+static void oacp_read_proc_execute(struct bt_gatt_ots *ots,
+				   struct bt_conn *conn)
+{
+	struct bt_gatt_ots_oacp_read_params *params =
+		&ots->cur_obj->state.read_op.oacp_params;
+
+	if (!ots->cur_obj) {
+		LOG_ERR("Invalid Current Object on OACP Read procedure");
+		return;
+	}
+
+	LOG_DBG("Executing Read procedure with offset: 0x%08X and "
+		"length: 0x%08X", params->offset, params->len);
+
+	if (ots->cb->obj_read) {
+		oacp_read_proc_cb(&ots->l2cap, conn);
+	} else {
+		ots->cur_obj->state.type = BT_GATT_OTS_OBJECT_IDLE_STATE;
+		LOG_ERR("OTS Read operation failed: "
+			"there is no OTS Read callback");
+	}
+}
+
+static void oacp_ind_cb(struct bt_conn *conn,
+			const struct bt_gatt_attr *attr,
+			uint8_t err)
+{
+	struct bt_gatt_ots *ots = (struct bt_gatt_ots *) attr->user_data;
+
+	LOG_DBG("Received OACP Indication ACK with status: 0x%04X", err);
+
+	switch (ots->cur_obj->state.type) {
+	case BT_GATT_OTS_OBJECT_READ_OP_STATE:
+		oacp_read_proc_execute(ots, conn);
+		break;
+	default:
+		LOG_ERR("Unsupported OTS state: %d", ots->cur_obj->state.type);
+		break;
+	}
+}
+
+static int oacp_ind_send(const struct bt_gatt_attr *oacp_attr,
+			 enum bt_gatt_ots_oacp_proc_type req_op_code,
+			 enum bt_gatt_ots_oacp_res_code oacp_status)
+{
+	uint8_t oacp_res[OACP_RES_MAX_SIZE];
+	uint16_t oacp_res_len = 0;
+	struct bt_gatt_ots *ots = (struct bt_gatt_ots *) oacp_attr->user_data;
+
+	/* Encode OACP Response */
+	oacp_res[oacp_res_len++] = BT_GATT_OTS_OACP_PROC_RESP;
+	oacp_res[oacp_res_len++] = req_op_code;
+	oacp_res[oacp_res_len++] = oacp_status;
```

Comment:
Maybe have a struct for this?

Reply:
I would rather not have one, it's one-time thing